File: janex-format/src/checksum.rs
// Copyright (c) 2026 Glavo
// SPDX-License-Identifier: MPL-2.0

//! Standard digest encodings and streaming checksum verification.

use crate::{Error, ErrorKind, Result, error::invalid};
use sha2::{Digest, Sha256, Sha512};
use std::io::Read;

/// A supported checksum algorithm and its on-disk identifier.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
#[repr(u8)]
pub enum Algorithm {
    /// XXH3-64 with the default secret, seed zero, and a big-endian digest; not cryptographically secure.
    Xxh3_64 = 0x11,
    /// XXH3-128 with the default secret, seed zero, and a big-endian digest; not cryptographically secure.
    Xxh3_128 = 0x12,
    /// SHA-256 with a 32-byte digest.
    Sha256 = 0x21,
    /// SHA-512 with a 64-byte digest.
    Sha512 = 0x22,
    /// SM3 with a 32-byte digest.
    Sm3 = 0x31,
}

impl Algorithm {
    /// Returns the digest length in bytes.
    pub fn digest_length(self) -> usize {
        match self {
            Self::Xxh3_64 => 8,
            Self::Xxh3_128 => 16,
            Self::Sha256 | Self::Sm3 => 32,
            Self::Sha512 => 64,
        }
    }

    /// Returns whether this algorithm may authenticate section or external-region bytes.
    pub fn is_secure(self) -> bool {
        matches!(self, Self::Sha256 | Self::Sha512 | Self::Sm3)
    }

    /// Decodes a supported algorithm identifier; zero is invalid, other unknown IDs unsupported.
    pub fn from_id(id: u8) -> Result<Self> {
        match id {
            0 => Err(invalid("checksum algorithm zero is reserved")),
            0x11 => Ok(Self::Xxh3_64),
            0x12 => Ok(Self::Xxh3_128),
            0x21 => Ok(Self::Sha256),
            0x22 => Ok(Self::Sha512),
            0x31 => Ok(Self::Sm3),
            _ => Err(Error::new(
                ErrorKind::Unsupported,
                format!("checksum algorithm {id}"),
            )),
        }
    }
}

/// An algorithm identifier followed by its standard digest bytes.
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct Checksum {
    /// The supported algorithm.
    algorithm: Algorithm,
    /// The exact standard digest representation.
    digest: Vec<u8>,
}

impl Checksum {
    /// Reads one complete checksum field, validating its algorithm and exact length.
    pub fn decode(bytes: &[u8]) -> Result<Self> {
        let (&id, digest) = bytes
            .split_first()
            .ok_or_else(|| invalid("empty checksum"))?;
        let algorithm = Algorithm::from_id(id)?;
        if digest.len() != algorithm.digest_length() {
            return Err(invalid("incorrect checksum length"));
        }
        Ok(Self {
            algorithm,
            digest: digest.to_vec(),
        })
    }

    /// Returns the algorithm.
    pub fn algorithm(&self) -> Algorithm {
        self.algorithm
    }

    /// Borrows the digest bytes without the algorithm identifier.
    pub fn digest(&self) -> &[u8] {
        &self.digest
    }

    /// Encodes the complete checksum, suitable for a binary field or CBOR byte string.
    pub fn encode(&self) -> Vec<u8> {
        let mut bytes = vec![self.algorithm as u8];
        bytes.extend_from_slice(&self.digest);
        bytes
    }

    /// Computes a checksum over all remaining bytes in a reader, consuming it through EOF.
    ///
    /// An I/O failure may have consumed a prefix of the input. No partial checksum is returned.
    pub fn compute(algorithm: Algorithm, mut reader: impl Read) -> Result<Self> {
        let digest = match algorithm {
            Algorithm::Xxh3_64 | Algorithm::Xxh3_128 => {
                let mut state = xxhash_rust::xxh3::Xxh3::new();
                read_chunks(&mut reader, |bytes| state.update(bytes))?;
                if algorithm == Algorithm::Xxh3_64 {
                    state.digest().to_be_bytes().to_vec()
                } else {
                    state.digest128().to_be_bytes().to_vec()
                }
            }
            Algorithm::Sha256 => hash::<Sha256>(&mut reader)?,
            Algorithm::Sha512 => hash::<Sha512>(&mut reader)?,
            Algorithm::Sm3 => hash::<sm3::Sm3>(&mut reader)?,
        };
        Ok(Self { algorithm, digest })
    }

    /// Verifies all remaining reader bytes, returning a verification error on mismatch.
    ///
    /// This checks content integrity and does not establish publisher identity or trust.
    pub fn verify(&self, reader: impl Read) -> Result<()> {
        if Self::compute(self.algorithm, reader)?.digest != self.digest {
            return Err(Error::new(ErrorKind::Verification, "checksum mismatch"));
        }
        Ok(())
    }
}

/// Hashes the remaining stream using a RustCrypto digest implementation.
fn hash<D: Digest + Default>(reader: &mut impl Read) -> Result<Vec<u8>> {
    let mut state = D::default();
    read_chunks(reader, |bytes| state.update(bytes))?;
    Ok(state.finalize().to_vec())
}

/// Passes bounded chunks to a consumer, retrying interrupted reads.
fn read_chunks(reader: &mut impl Read, mut consume: impl FnMut(&[u8])) -> Result<()> {
    let mut buffer = [0; 64 * 1024];
    loop {
        match reader.read(&mut buffer) {
            Ok(0) => return Ok(()),
            Ok(count) => consume(&buffer[..count]),
            Err(error) if error.kind() == std::io::ErrorKind::Interrupted => {}
            Err(error) => return Err(error.into()),
        }
    }
}

File: janex-reader/src/main/java/org/janex/format/Checksum.java
// Copyright (c) 2026 Glavo
// SPDX-License-Identifier: MPL-2.0

package org.janex.format;

import java.io.IOException;
import java.io.InputStream;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.Arrays;
import java.util.Objects;

/// An immutable Janex algorithm identifier and its canonical digest bytes.
///
/// Successful verification establishes content integrity, not publisher identity.
public final class Checksum {
    /// Supported algorithms with the identifiers and byte representations defined by Janex 0.1.
    public enum Algorithm {
        /// Seed-zero XXH3-64 with the default secret and a big-endian digest; not cryptographically secure.
        XXH3_64(0x11, 8, false),
        /// Seed-zero XXH3-128 with the default secret and the high half first; not cryptographically secure.
        XXH3_128(0x12, 16, false),
        /// SHA-256 in its standard 32-byte representation.
        SHA256(0x21, 32, true),
        /// SHA-512 in its standard 64-byte representation.
        SHA512(0x22, 64, true),
        /// SM3 in its standard 32-byte representation.
        SM3(0x31, 32, true);

        /// On-disk algorithm identifier.
        private final int id;
        /// Required digest length in bytes.
        private final int length;
        /// Whether the algorithm is permitted for cryptographic content authentication.
        private final boolean secure;

        /// Retains one fixed algorithm definition.
        Algorithm(int id, int length, boolean secure) {
            this.id = id;
            this.length = length;
            this.secure = secure;
        }

        /// Returns the unsigned byte identifier stored in a checksum field.
        public int id() {
            return id;
        }

        /// Returns the required digest length, excluding the algorithm identifier.
        public int digestLength() {
            return length;
        }

        /// Returns whether this algorithm may authenticate section or external-region bytes.
        public boolean isSecure() {
            return secure;
        }

        /// Creates independent, resettable streaming state with bounded internal buffering.
        ///
        /// The state follows [MessageDigest]'s update and digest contracts and is not thread-safe.
        /// XXH3 always uses seed zero and the default secret. SM3 does not require a JCA provider.
        ///
        /// @return a new digest whose output uses this algorithm's canonical byte representation
        /// @throws IOException if a required platform SHA implementation is unavailable
        public MessageDigest newDigest() throws IOException {
            switch (this) {
                case XXH3_64:
                    return new Xxh3(false);
                case XXH3_128:
                    return new Xxh3(true);
                case SM3:
                    return new Sm3();
                default:
                    try {
                        return MessageDigest.getInstance(this == SHA256 ? "SHA-256" : "SHA-512");
                    } catch (NoSuchAlgorithmException failure) {
                        throw new IOException("Required digest is unavailable: " + this, failure);
                    }
            }
        }

        /// Decodes one supported algorithm identifier.
        ///
        /// @param id unsigned byte value from a checksum field
        /// @return the corresponding supported algorithm
        /// @throws IOException if the identifier is reserved, unknown, or outside the unsigned byte range
        public static Algorithm fromId(int id) throws IOException {
            for (Algorithm algorithm : values()) {
                if (algorithm.id == id) {
                    return algorithm;
                }
            }
            throw new IOException(id == 0 ? "Checksum algorithm zero is reserved" : "Unsupported checksum algorithm: " + id);
        }
    }

    /// Supported checksum algorithm.
    private final Algorithm algorithm;
    /// Owned digest in canonical byte order.
    private final byte[] digest;

    /// Retains an already validated, owned digest.
    private Checksum(Algorithm algorithm, byte[] digest) {
        this.algorithm = algorithm;
        this.digest = digest;
    }

    /// Decodes a complete binary or CBOR checksum field without retaining its array.
    ///
    /// @param bytes algorithm byte followed by its exact canonical digest
    /// @return an immutable checksum
    /// @throws IOException if the field is empty, its algorithm is unsupported, or its length is incorrect
    /// @throws NullPointerException if bytes is null
    public static Checksum decode(byte[] bytes) throws IOException {
        Objects.requireNonNull(bytes);
        Input.require(bytes.length != 0, "Empty checksum");
        Algorithm algorithm = Algorithm.fromId(bytes[0] & 255);
        Input.require(bytes.length == algorithm.length + 1, "Incorrect checksum length");
        return new Checksum(algorithm, Arrays.copyOfRange(bytes, 1, bytes.length));
    }

    /// Returns the checksum algorithm.
    public Algorithm algorithm() {
        return algorithm;
    }

    /// Returns a new array containing the canonical digest without its algorithm byte.
    public byte[] digest() {
        return digest.clone();
    }

    /// Returns a new array containing the algorithm byte followed by the canonical digest.
    public byte[] encode() {
        byte[] encoded = new byte[digest.length + 1];
        encoded[0] = (byte) algorithm.id;
        System.arraycopy(digest, 0, encoded, 1, digest.length);
        return encoded;
    }

    /// Computes a checksum over an unchanged byte array without retaining it.
    ///
    /// @param algorithm supported checksum algorithm
    /// @param bytes complete input, not modified
    /// @return the computed checksum
    /// @throws IOException if the required platform digest is unavailable
    /// @throws NullPointerException if either argument is null
    public static Checksum compute(Algorithm algorithm, byte[] bytes) throws IOException {
        return new Checksum(algorithm, algorithm.newDigest().digest(Objects.requireNonNull(bytes)));
    }

    /// Computes a checksum over all remaining bytes, consuming the stream through EOF without closing it.
    ///
    /// An I/O failure may consume a prefix; no partial checksum is returned. The input is read with
    /// a fixed-size buffer, independently of the reader's buffered-value limits.
    ///
    /// @param algorithm supported checksum algorithm
    /// @param input stream owned by the caller
    /// @return the computed checksum
    /// @throws IOException if reading fails or the required platform digest is unavailable
    /// @throws NullPointerException if either argument is null
    public static Checksum compute(Algorithm algorithm, InputStream input) throws IOException {
        Objects.requireNonNull(input);
        MessageDigest digest = algorithm.newDigest();
        byte[] buffer = new byte[32768];
        for (int count; (count = input.read(buffer)) != -1;) {
            if (count == 0) {
                int value = input.read();
                if (value == -1) {
                    break;
                }
                digest.update((byte) value);
            } else {
                digest.update(buffer, 0, count);
            }
        }
        return new Checksum(algorithm, digest.digest());
    }

    /// Verifies an unchanged byte array against this checksum.
    ///
    /// @param bytes complete input, not modified
    /// @throws IOException if the checksum differs or its required platform digest is unavailable
    /// @throws NullPointerException if bytes is null
    public void verify(byte[] bytes) throws IOException {
        verifyDigest(compute(algorithm, bytes).digest);
    }

    /// Verifies all remaining stream bytes without closing the stream.
    ///
    /// Consumption and failure behavior are as specified by [#compute(Algorithm, InputStream)].
    ///
    /// @param input stream owned by the caller
    /// @throws IOException if reading fails, the checksum differs, or its platform digest is unavailable
    /// @throws NullPointerException if input is null
    public void verify(InputStream input) throws IOException {
        verifyDigest(compute(algorithm, input).digest);
    }

    /// Checks canonical digest bytes already calculated by a streaming caller.
    void verifyDigest(byte[] actual) throws IOException {
        Input.require(MessageDigest.isEqual(digest, actual), "Janex checksum mismatch");
    }
}
